docs/proposals/20200602-machine-deletion-phase-hooks.md

```diff
@@ -0,0 +1,301 @@
+linear progress of a machine’s lifecycle after a machine has been marked for
+deletion.  These annotations are optional and may be applied during machine
+creation, sometime after machine creation by a user, or sometime after machine
+creation by another controller or application.
+
+## Motivation
+
+Allow custom and 3rd party components to easily interact with a machine or
+related resources while that machine's reconciliation is temporarily paused.
+This pause in reconciliation will allow these custom components to take action
+after a machine has been marked for deletion, but prior to the machine being
+drained and/or associated instance terminated.
+
+### Goals
+
+- Define an initial set of hook points for the deletion phase.
+- Define an initial set and form of related annotations.
+- Define basic expectations for a controller or process that responds to a
+lifecycle hook.
+
+
+### Non-Goals/Future Work
+
+- Create an exhaustive list of hooks; we can add more over time.
+- Create new machine phases
+- Dictate implementation of controllers that respond to the hooks.
+- Implement ordering in the machine-controller.
+- Require anyone use these hooks for normal machine operations, these are
+strictly optional and for custom integrations only.
+
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+
+### User Stories
+#### Story 1
+(pre-term) As an operator, I would like to have the ability to perform
+different actions between the time a machine is marked deleted in the api and
+the time the machine is deleted from the cloud.
+
+For example, when replacing a control plane machine, ensure a new control
+plane machine has been successfully created and joined to the cluster before
+removing the instance of the deleted machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.  This will also prevent
+prolonged periods of having one fewer control plane host in the event the
+replacement instance does not come up in a timely manner.
+
+#### Story 2
+(pre-drain) As an operator, I want the ability to utilize my own draining
+controller instead of the logic built into the machine-controller.  This will
+allow me better flexibility and control over the lifecycle of workloads on each
+node.
+
+#### Story 3
+(pre-drain) As an operator, when I am deleting a control plane machine for
+maintenance reasons, I want to ensure my existing control plane machine is not
+drained until after my replacement comes online.  This will prevent protracted
+periods of a missing control plane member if the replacement machine cannot be
```

Comment:
I'm fine with changing the wording, but I don't see how this is at odds with MHC.  MHC can delete a machine, as normal.  We just prevent the instance from being drained/terminated until the replacement comes online.  To the MHC, this just looks like it's taking a long time to drain the machine, it doesn't need to do anything different.

I think when we were discussing the modifications to MHC for the KCP previously, those bits might be a temporary fix until something like this merges.

Reply:
My concern is that the work that is in-flight (https://github.com/kubernetes-sigs/cluster-api/pull/3185) will block creation of the new Machine until the old Machine has been fully deleted. This is actually needed to avoid cases where scaling up the etcd cluster prior to fully removing the machine being remediated could lead to a loss of quorum.

For example, given a 3 replica Control Plane:
- 1 machine fails MHC in a way that also impacts the etcd member, we are now at 2/3 healthy etcd members and still have quorum
- If we attempt to scale up the Control Plane, part way through the bootstrapping of the 4th replica, the new member will be added to etcd, at this point quorum is lost (2/4 members healthy)
- At this point one of two things can happen:
  - The new etcd member finishes syncing, quorum is restored, and the 4th replica finishes bootstrapping. Cluster wide impact is relatively minimal
  - The etcd member fails to sync, quorum loss remains. Cluster wide services impacted due to read-only datastore.

Reply:
> If we attempt to scale up the Control Plane, part way through the bootstrapping of the 4th replica, the new member will be added to etcd, at this point quorum is lost (2/4 members healthy)

I have thought about this at great length in the context of what we're doing in OpenShift.

The answer I came up with was divorcing etcd from automatically being enrolled as a member of the cluster upon control plane scale up.  We can install the binaries, we can configure it, but just don't add the member to the cluster until some smarter thing elects to do that.  So, you could have 6 control plane hosts, and only 3 active etcd members.  The others could be 'warm' so you can quickly remove/add the hosts you want.  The etcd controller would need to have some idea about fault domains and machine/node statuses, but that is the best way IMO.

This also lets us scale etcd independently of scaling control plane hosts.  Want to have a single control plane host with a single etcd member?  Easy, scale etcd down to one, then scale down your control plane hosts.  etcd controller will ensure that etcd lands in the right spot.  Want to scale up to 5 etcd hosts?  Fine, scale up the control plane, scale up etcd.

It solves so many problems.  I'm not sure exactly how it fits into the existing architecture, but this is the only reasonable setup I can see.

In any case, what we're defining in this CAEP is optional behavior of 3rd party components.  Since cluster-api bits are supposed to be reusable (that is still the case, right?), this use case makes sense IMO.  If it's creating confusion as you say, I can remove it.

Reply:
There are probably more nuanced stories we can tease out specifically around control plane / etcd membership and deletion hooks. To move forward with this proposal, though, let's remove this story as you suggested.